lib/gateway.js

```diff
@@ -0,0 +1,250 @@
+  }
+}
+
+function buildRequest (opts) {
+  const agents = {
+    'http:': new http.Agent(agentOption(opts)),
+    'https:': new https.Agent(agentOption(opts))
+  }
+
+  const requests = {
+    'http:': http,
+    'https:': https
+  }
+
+  function close () {
+    agents['http:'].destroy()
+    agents['https:'].destroy()
+  }
+
+  function request (opts, done) {
+    // console.log('sending request', opts.method, opts.url.port, opts.url.pathname, opts.url.hostname)
+    const req = requests[opts.url.protocol].request({
+      method: opts.method,
+      port: opts.url.port,
+      path: opts.url.pathname + (opts.qs || ''),
+      hostname: opts.url.hostname,
+      headers: opts.headers,
+      agent: agents[opts.url.protocol]
+    })
+    req.on('error', done)
+    req.on('response', res => {
+      done(null, { statusCode: res.statusCode, headers: res.headers, stream: res })
+    })
+    end(req, opts.body, done)
+  }
+
+  return {
+    request,
+    close
+  }
+}
+
+function createTypeMap (schemaDefinition) {
+  const parsedSchema = parse(schemaDefinition)
+  const typeMap = {}
+
+  for (const definition of parsedSchema.definitions) {
+    if (isTypeDefinitionNode(definition)) {
+      typeMap[definition.name.value] = new Set(definition.fields.map(field => field.name.value))
+    }
+  }
+
+  return typeMap
+}
+
+async function buildServiceMap (services) {
+  const serviceMap = {}
+
+  for (const service of services) {
+    const { name, ...opts } = service
+    const { request, close } = buildRequest(opts)
+    const url = new URL(opts.url)
+
+    const serviceConfig = {
+      request: function (opts) {
+        return new Promise((resolve, reject) => {
+          request({
+            url,
+            method: opts.method || 'POST',
+            body: opts.body,
+            headers: {
+              'content-type': 'application/json',
+              ...opts.headers
+            }
+          }, (err, response) => {
+            if (err) {
+              return reject(err)
+            }
+
+            if (response.statusCode === 200) {
+              response.stream.on('data', data => resolve({
+                statusCode: response.statusCode,
+                json: JSON.parse(data.toString())
+              }))
+            }
+          })
+        })
+      },
+      close,
+      init: async () => {
+        const response = await serviceConfig.request({
+          method: 'POST',
+          body: JSON.stringify({
+            query: `
+            query ServiceInfo {
+              _service {
+                sdl
+              }
+            }
+            `
+          })
+        })
+
+        const schemaDefinition = response.json.data._service.sdl
+
+        const typeMap = createTypeMap(schemaDefinition)
+
+        return {
+          schemaDefinition,
+          typeMap
+        }
+      }
+    }
+
+    serviceMap[service.name] = serviceConfig
+  }
+
+  await Promise.all(services.map(service => {
```

Comment:
This is not going to work well if there are many services. Use http://npm.im/p-map with a max concurrency of 8.